File: airbyte-server/src/main/java/io/airbyte/server/apis/SignalApiController.java
/*
 * Copyright (c) 2020-2025 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.server.apis;

import static io.airbyte.commons.auth.AuthRoleConstants.ADMIN;

import io.airbyte.api.generated.SignalApi;
import io.airbyte.api.model.generated.SignalInput;
import io.airbyte.commons.server.handlers.SignalHandler;
import io.airbyte.commons.server.scheduling.AirbyteTaskExecutors;
import io.micronaut.http.annotation.Body;
import io.micronaut.http.annotation.Controller;
import io.micronaut.http.annotation.Post;
import io.micronaut.scheduling.annotation.ExecuteOn;
import io.micronaut.security.annotation.Secured;
import io.micronaut.security.rules.SecurityRule;

@Controller("/api/v1/signal")
@Secured(SecurityRule.IS_AUTHENTICATED)
public class SignalApiController implements SignalApi {

  final SignalHandler signalHandler;

  public SignalApiController(SignalHandler signalHandler) {
    this.signalHandler = signalHandler;
  }

  @Override
  @Post
  @ExecuteOn(AirbyteTaskExecutors.IO)
  @Secured({ADMIN})
  public void signal(@Body SignalInput signalInput) {
    ApiHelper.execute(() -> {
      final io.airbyte.config.SignalInput internalSignalInput = new io.airbyte.config.SignalInput(
          signalInput.getWorkflowType(),
          signalInput.getWorkflowId());
      signalHandler.signal(internalSignalInput);
      return null;
    });
  }

}

File: airbyte-server/src/main/kotlin/io/airbyte/server/apis/controllers/SignalApiController.kt
/*
 * Copyright (c) 2020-2025 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.server.apis.controllers

import io.airbyte.api.generated.SignalApi
import io.airbyte.api.model.generated.SignalInput
import io.airbyte.commons.auth.roles.AuthRoleConstants
import io.airbyte.commons.server.handlers.SignalHandler
import io.airbyte.commons.server.scheduling.AirbyteTaskExecutors
import io.airbyte.server.apis.execute
import io.micronaut.http.annotation.Body
import io.micronaut.http.annotation.Controller
import io.micronaut.http.annotation.Post
import io.micronaut.scheduling.annotation.ExecuteOn
import io.micronaut.security.annotation.Secured
import io.micronaut.security.rules.SecurityRule

@Controller("/api/v1/signal")
@Secured(SecurityRule.IS_AUTHENTICATED)
class SignalApiController(
  val signalHandler: SignalHandler,
) : SignalApi {
  @Post
  @ExecuteOn(AirbyteTaskExecutors.IO)
  @Secured(AuthRoleConstants.ADMIN)
  override fun signal(
    @Body signalInput: SignalInput,
  ) {
    execute<Any?> {
      val internalSignalInput =
        io.airbyte.config.SignalInput(
          signalInput.workflowType,
          signalInput.workflowId,
        )
      signalHandler.signal(internalSignalInput)
      null
    }
  }
}
